### PR TITLE
Fix Aggregate:accumulatorAlignmentSize()

### PR DESCRIPTION
**Pull request by Yuhta**
Summary:
There is a problem that `Aggregate:accumulatorAlignmentSize` only applies to the offset inside a group, not the group address itself.  This causes `Aggregate::value` returns unaligned address and crashes for example with `int128_t`.

This change fixes this by aligning the whole group memory, including the allocation run, the normalized key, and the group content itself.

Differential Revision: D44015635

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@Yuhta It would be nice to update documentation at https://facebookincubator.github.io/velox/develop/aggregate-functions.html#memory-layout

File: velox/common/memory/AllocationPool.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ class AllocationPool {
 
   void clear();
 
-  char* allocateFixed(uint64_t bytes);
+  char* allocateFixed(uint64_t bytes, int alignment = 1);
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.h:38]: int -> int32_t

Would you document this method and add a test? Consider extracting this change into a separate PR.

File: velox/exec/GroupingSet.cpp [hunk 2]
```diff
@@ -289,9 +290,14 @@ void GroupingSet::initializeGlobalAggregation() {
         rowSizeOffset);
     offset += aggregate->accumulatorFixedWidthSize();
     ++nullOffset;
+    VELOX_CHECK_EQ(
+        __builtin_popcount(aggregate->accumulatorAlignmentSize()),
+        1,
+        "Alignment can only be power of 2");
+    alignment = std::max(alignment, aggregate->accumulatorAlignmentSize());
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:297]: Perhaps, update the comment on L270.

File: velox/exec/GroupingSet.cpp [hunk 3]
```diff
@@ -277,6 +277,7 @@ void GroupingSet::initializeGlobalAggregation() {
   int32_t rowSizeOffset = bits::nbytes(aggregates_.size());
   int32_t offset = rowSizeOffset + sizeof(int32_t);
   int32_t nullOffset = 0;
+  int alignment = 1;
```
Comment (mbasmanova) [inline velox/exec/GroupingSet.cpp:280]: int32_t

File: velox/exec/RowContainer.h [hunk 4]
```diff
@@ -367,10 +363,11 @@ class RowContainer {
         auto row = iter->rowOffset;
         while (row + rowSize <= limit) {
           rows[count++] = data + row +
-              (iter->normalizedKeysLeft > 0 ? sizeof(normalized_key_t) : 0);
+              (iter->normalizedKeysLeft > 0 ? normalizedKeySize() : 0);
           row += rowSize;
+          auto totalBytes2 = totalBytes + rowSize;
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:368]: > totalBytes2

Would it be possible to come up with a more descriptive name? Should we add a test case for the changes in RowContainer?

File: velox/exec/RowContainer.cpp [hunk 5]
```diff
@@ -109,29 +109,24 @@ RowContainer::RowContainer(
...
-    offset += aggregate->accumulatorFixedWidthSize();
     nullOffsets_.push_back(nullOffset);
     ++nullOffset;
     isVariableWidth |= !aggregate->isFixedSize();
     usesExternalMemory_ |= aggregate->accumulatorUsesExternalMemory();
+    VELOX_CHECK_EQ(
+        __builtin_popcount(aggregate->accumulatorAlignmentSize()),
+        1,
+        "Alignment can only be power of 2");
+    alignment_ = std::max(alignment_, aggregate->accumulatorAlignmentSize());
```
Comment (mbasmanova) [inline velox/exec/RowContainer.cpp:120]: This code repeats in two places. Consider introducing a helper function to avoid copy-paste.

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Some initial comments.

**Comment by majetideepak** [comment]
@Yuhta can you cherry-pick this PR https://github.com/facebookincubator/velox/pull/4154/files on top of your changes and see if the test crashes?

**Comment by majetideepak** [comment]
The linux-adapters CI test should pass with the above PR and your changes.

**Comment by isadikov** [comment]
I am trying to verify the fix locally but I am getting this error when building the branch: 
```
FAILED: velox/functions/prestosql/aggregates/tests/velox_aggregates_test 
: && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -mcpu=apple-m1+crc -std=c++17 -fvisibility=hidden -fvisibility=hidden -fvisibility-inlines-hidden -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -mmacosx-version-min=12.5 -Wl,-search_paths_first -Wl,-headerpad_max_install_names -Wl,-flat_namespace velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/HiveConfig.cpp.o velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/HiveConnector.cpp.o velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/HiveDataSink.cpp.o velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/HivePartitionUtil.cpp.o velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/FileHandle.cpp.o velox/connectors/hive/CMakeFiles/velox_hive_connector.dir/PartitionIdGenerator.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/AggregationFunctionRegTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ApproxDistinctTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ApproxMostFrequentTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ApproxPercentileTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ArbitraryTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ArrayAggTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/AverageAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/BitwiseAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/BoolAndOrTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ChecksumAggregateTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/CountAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/CountIfAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/CovarianceAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/HistogramTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/Main.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/MinMaxByAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/MinMaxTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/PrestoHasherTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/SumTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/MapAggTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/MapUnionAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/ValueListTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/VarianceAggregationTest.cpp.o velox/functions/prestosql/aggregates/tests/CMakeFiles/velox_aggregates_test.dir/MaxSizeForStatsTest.cpp.o -o velox/functions/prestosql/aggregates/tests/velox_aggregates_test  -Wl,-rpath,/Users/ivansadikov/homebrew/lib  velox/functions/prestosql/aggregates/libvelox_aggregates.a  velox/functions/prestosql/aggregates/tests/libvelox_aggregates_test_lib.a  velox/core/libvelox_core.a  velox/dwio/common/tests/utils/libvelox_dwio_common_test_utils.a  velox/exec/libvelox_exec.a  velox/exec/tests/utils/libvelox_exec_test_lib.a  velox/common/file/libvelox_file.a  velox/functions/prestosql/tests/utils/libvelox_functions_test_lib.a  velox/functions/prestosql/registration/libvelox_functions_prestosql.a  velox/functions/lib/libvelox_functions_lib.a  velox/type/libvelox_type.a  velox/vector/fuzzer/libvelox_vector_fuzzer.a  lib/libgtestd.a  lib/libgtest_maind.a  velox/exec/tests/utils/libvelox_exec_test_lib.a  velox/functions/prestosql/aggregates/libvelox_aggregates.a  velox/dwio/common/tests/utils/libvelox_dwio_common_test_utils.a  velox/functions/prestosql/registration/libvelox_functions_prestosql.a  velox/functions/prestosql/libvelox_functions_prestosql_impl.a  velox/functions/lib/libvelox_functions_lib.a  /usr/local/lib/libre2.a  velox/common/hyperloglog/libvelox_common_hyperloglog.a  velox/functions/prestosql/json/libvelox_functions_json.a  velox/external/md5/libmd5.a  velox/functions/prestosql/types/libvelox_presto_types.a  velox/functions/lib/libvelox_is_null_functions.a  velox/serializers/libvelox_presto_serializer.a  velox/dwio/type/fbhive/libvelox_dwio_type_fbhive.a  velox/exec/tests/utils/libvelox_temp_path.a  velox/dwio/dwrf/reader/libvelox_dwio_dwrf_reader.a  velox/dwio/dwrf/writer/libvelox_dwio_dwrf_writer.a  velox/dwio/dwrf/common/libvelox_dwio_dwrf_common.a  velox/dwio/common/libvelox_dwio_common.a  velox/dwio/common/exception/libvelox_dwio_common_exception.a  velox/dwio/common/encryption/libvelox_dwio_common_encryption.a  velox/common/caching/libvelox_caching.a  velox/dwio/common/compression/libvelox_dwio_common_compression.a  velox/dwio/dwrf/utils/libvelox_dwio_dwrf_utils.a  velox/dwio/dwrf/proto/libvelox_dwio_dwrf_proto.a  /opt/homebrew/lib/libprotobuf.dylib  /opt/homebrew/lib/liblz4.dylib  /opt/homebrew/lib/liblzo2.dylib  /opt/homebrew/lib/libzstd.dylib  velox/tpch/gen/libvelox_tpch_gen.a  velox/parse/libvelox_parse_parser.a  velox/duckdb/conversion/libvelox_duckdb_parser.a  velox/duckdb/conversion/libvelox_duckdb_conversion.a  velox/external/duckdb/tpch/libtpch_extension.a  velox/external/duckdb/tpch/dbgen/libdbgen.a  velox/parse/libvelox_parse_expression.a  velox/parse/libvelox_parse_utils.a  velox/functions/libvelox_function_registry.a  velox/external/duckdb/libduckdb.a  velox/vector/tests/utils/libvelox_vector_test_lib.a  velox/exec/libvelox_exec.a  velox/codegen/libvelox_codegen.a  velox/exec/libvelox_exec.a  velox/codegen/libvelox_codegen.a  velox/common/file/libvelox_file.a  velox/connectors/libvelox_connector.a  velox/vector/arrow/libvelox_arrow_bridge.a  velox/expression/libvelox_expression.a  velox/core/libvelox_core.a  velox/core/libvelox_config.a  velox/functions/lib/libvelox_functions_util.a  velox/expression/libvelox_expression_functions.a  velox/expression/type_calculation/libvelox_type_calculation.a  velox/vector/libvelox_vector.a  velox/type/libvelox_type.a  velox/common/serialization/libvelox_serialization.a  velox/type/tz/libvelox_type_tz.a  velox/external/date/libvelox_external_date.a  velox/common/time/libvelox_time.a  velox/common/encode/libvelox_encode.a  velox/buffer/libvelox_buffer.a  velox/common/memory/libvelox_memory.a  velox/common/testutil/libvelox_test_util.a  velox/common/base/libvelox_common_base.a  velox/common/base/libvelox_exception.a  velox/common/process/libvelox_process.a  /usr/local/lib/libfolly.a  /usr/local/lib/libfmt.a  /Users/ivansadikov/homebrew/lib/libgflags.2.2.2.dylib  /Users/ivansadikov/homebrew/lib/libglog.dylib  /Users/ivansadikov/homebrew/lib/libevent.dylib  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/lib/libz.tbd  /Users/ivansadikov/homebrew/opt/openssl@1.1/lib/libssl.dylib  /Users/ivansadikov/homebrew/opt/openssl@1.1/lib/libcrypto.dylib  /Users/ivansadikov/homebrew/lib/liblz4.dylib  /Users/ivansadikov/homebrew/lib/libzstd.dylib  /Users/ivansadikov/homebrew/lib/libsnappy.dylib  /Users/ivansadikov/homebrew/lib/libsodium.dylib  -lc++abi  /Users/ivansadikov/homebrew/lib/libboost_context-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_filesystem-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_atomic-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_program_options-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_regex-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_system-mt.dylib  /Users/ivansadikov/homebrew/lib/libboost_thread-mt.dylib  /usr/local/lib/libdouble-conversion.a  /opt/homebrew/lib/libevent.dylib  /opt/homebrew/lib/libsnappy.dylib  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk/usr/lib/libbz2.tbd  /opt/homebrew/lib/libglog.dylib  /Users/ivansadikov/homebrew/lib/libgflags.2.2.2.dylib  lib/libgtestd.a && :
Undefined symbols for architecture arm64:
  "facebook::velox::AllocationPool::allocateFixed(unsigned long long)", referenced from:
      facebook::velox::exec::GroupingSet::initializeGlobalAggregation() in libvelox_exec.a(GroupingSet.cpp.o)
     (maybe you meant: facebook::velox::AllocationPool::allocateFixed(unsigned long long, int))
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Could you check this? Thanks.

**Comment by Yuhta** [comment]
@isadikov Maybe do a full clean and rebuild?  Looks like you are referencing an old function signature.

File: velox/docs/develop/aggregate-functions.rst [hunk 6]
```diff
@@ -92,12 +92,18 @@ Fixed-width part of the accumulator is stored in the row. Variable-width
...
-accumulators.
-
-.. image:: images/aggregation-layout.png
-  :width: 600
+A row is a contiguous byte buffer.  Data is stored in the following order:
+1. Null flags (1 bit per item) for 
+    1. Keys (only if nullable)
+    2. Accumulators
+    3. Dependants
+2. Has-probed flag (1 bit)
```
Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:100]: Perhaps, clarify that this flag is used only for joins.

File: velox/docs/develop/aggregate-functions.rst [hunk 7]
```diff
@@ -92,12 +92,18 @@ Fixed-width part of the accumulator is stored in the row. Variable-width
...
+3. Free flag (1 bit)
+4. Keys
+5. Accumulators
```
Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:103]: Perhaps, clarify that this is only fixed-width part of the accumulators.

File: velox/docs/develop/aggregate-functions.rst [hunk 8]
```diff
@@ -92,12 +92,18 @@ Fixed-width part of the accumulator is stored in the row. Variable-width
...
+6. Dependants
```
Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:104]: Perhaps, clarify that this is used only for joins.

File: velox/docs/develop/aggregate-functions.rst [hunk 9]
```diff
@@ -92,12 +92,18 @@ Fixed-width part of the accumulator is stored in the row. Variable-width
...
+7. Variable size (32 bit)
```
Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:105]: Can you remind me what this is?

File: velox/docs/develop/aggregate-functions.rst [hunk 10]
```diff
@@ -92,12 +92,18 @@ Fixed-width part of the accumulator is stored in the row. Variable-width
 part (if exists) is allocated using :doc:`HashStringAllocator <arena>` and a pointer is
 stored in the fixed-width part.
 
-A row is a contiguous byte buffer. Given N aggregates, first N / 8 bytes
-store the null flags, one bit per aggregate, followed by fixed-width
-accumulators.
-
-.. image:: images/aggregation-layout.png
-  :width: 600
+A row is a contiguous byte buffer.  Data is stored in the following order:
```
Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:95]: It would be nice to clarify that sometimes there are gaps due to alignment. It would be helpful to explain where alignment boundaries are.

Comment (mbasmanova) [inline velox/docs/develop/aggregate-functions.rst:95]: I see velox/docs/develop/images/aggregation-layout.png in the PR, but can't see what has changed.

**Review by mbasmanova — COMMENTED** [review]
@Yuhta Thank you for updating the documentation. Some questions.

File: velox/common/memory/AllocationPool.cpp [hunk 11]
```diff
@@ -29,31 +29,49 @@ void AllocationPool::clear() {
   largeAllocations_.clear();
 }
 
-char* AllocationPool::allocateFixed(uint64_t bytes) {
-  VELOX_CHECK_GT(bytes, 0, "Cannot allocate zero bytes");
+namespace {
 
-  auto numPages = bits::roundUp(bytes, memory::AllocationTraits::kPageSize) /
+uint64_t getNumPages(uint64_t bytes) {
+  return bits::roundUp(bytes, memory::AllocationTraits::kPageSize) /
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.cpp:35]: nit: Would it make sense to make this a method of AllocationTraits?

```
memory::AllocationTraits::numPages(bytes)
```

File: velox/common/memory/AllocationPool.cpp [hunk 12]
```diff
@@ -29,31 +29,49 @@ void AllocationPool::clear() {
...
         res, "Unexpected nullptr for large contiguous allocation");
+    // Should be at page boundary and always aligned.
+    VELOX_CHECK_EQ(reinterpret_cast<uintptr_t>(res) % alignment, 0);
     return res;
   }
 
-  if (availableInRun() < bytes) {
-    newRunImpl(numPages);
+  if (availableInRun() == 0) {
+    newRunImpl(getNumPages(bytes + alignment - 1));
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.cpp:61]: Should we compute numPages at the beginning of this method, store it in a temp variable and reuse?

File: velox/common/memory/AllocationPool.cpp [hunk 13]
```diff
@@ -29,31 +29,49 @@ void AllocationPool::clear() {
...
+  } else if (auto alignedBytes =
+                 bytes + memory::alignmentPadding(firstFreeInRun(), alignment);
+             availableInRun() < alignedBytes) {
+    newRunImpl(getNumPages(alignedBytes));
   }
   auto run = currentRun();
+  currentOffset_ += memory::alignmentPadding(firstFreeInRun(), alignment);
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.cpp:68]: Should we compute memory::alignmentPadding(firstFreeInRun(), alignment) once and reuse?

File: velox/common/memory/AllocationPool.cpp [hunk 14]
```diff
@@ -29,31 +29,49 @@ void AllocationPool::clear() {
...
   uint64_t size = run.numBytes();
   VELOX_CHECK_LE(bytes + currentOffset_, size);
+  auto* res = run.data<char>() + currentOffset_;
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.cpp:71]: do not abbreviate: res -> result

File: velox/common/memory/AllocationPool.h [hunk 15]
```diff
@@ -35,7 +35,9 @@ class AllocationPool {
 
   void clear();
 
-  char* allocateFixed(uint64_t bytes);
+  // Allocate a buffer from this pool, optionally aligned.  The alignment can
+  // only be power of 2.
+  char* allocateFixed(uint64_t bytes, int32_t alignment = 1);
```
Comment (mbasmanova) [inline velox/common/memory/AllocationPool.h:40]: Would it be possible to extract this change into a separate PR to keeps PRs small and easy to review?

File: velox/exec/Aggregate.h [hunk 16]
```diff
@@ -47,12 +47,16 @@ class Aggregate {
...
 
-  /// Returns the alignment size of the accumulator.
-  /// Some types such as int128_t require aligned access.
+  /// Returns the alignment size of the accumulator.  Some types such as
+  /// int128_t require aligned access.  This value must be a power of 2.
   virtual int32_t accumulatorAlignmentSize() const {
     return 1;
   }
 
+  /// Combine the accumulator alignment requirement of this with the input
```
Comment (mbasmanova) [inline velox/exec/Aggregate.h:56]: Perhaps, explain what does it mean to "combine" two alignments.

File: velox/exec/Aggregate.cpp [hunk 17]
```diff
@@ -121,4 +121,13 @@ TypePtr Aggregate::intermediateType(
   VELOX_FAIL("Could not infer intermediate type for aggregate {}", name);
 }
 
+int32_t Aggregate::combineAlignment(int32_t otherAlign) const {
```
Comment (mbasmanova) [inline velox/exec/Aggregate.cpp:124]: naming: thisAlignment and otherAlignment (or alignment and other)

File: velox/exec/RowContainer.h [hunk 18]
```diff
@@ -1025,6 +1024,12 @@ class RowContainer {
   // Free any aggregates associated with the 'rows'.
   void freeAggregates(folly::Range<char**> rows);
 
+  int normalizedKeySize() const {
```
Comment (mbasmanova) [inline velox/exec/RowContainer.h:1027]: Any particular reason not to compute this once in the constructor?

File: velox/exec/tests/utils/SumNonPODAggregate.cpp [hunk 19]
```diff
@@ -127,10 +133,14 @@ class SumNonPODAggregate : public Aggregate {
       bool mayPushdown) override {
     addSingleGroupIntermediateResults(group, rows, args, mayPushdown);
   }
+
+ private:
+  int alignment_;
```
Comment (mbasmanova) [inline velox/exec/tests/utils/SumNonPODAggregate.cpp:138]: const int32_t

File: velox/functions/prestosql/aggregates/DecimalAggregate.h [hunk 20]
```diff
@@ -208,6 +208,7 @@ class DecimalAggregate : public exec::Aggregate {
       });
     } else {
       rows.applyToSelected([&](vector_size_t i) {
+        clearNull(groups[i]);
```
Comment (mbasmanova) [inline velox/functions/prestosql/aggregates/DecimalAggregate.h:211]: Is this a separate bug fix?

**Review by mbasmanova — APPROVED** [review]
@Yuhta Looks good % some comments.

File: velox/docs/develop/aggregate-functions.rst (hunk 9 above)
Comment (Yuhta) [reply velox/docs/develop/aggregate-functions.rst:105]: It's used to track the variable size of the row (for complex types etc.)

File: velox/docs/develop/aggregate-functions.rst (hunk 10 above)
Comment (Yuhta) [reply velox/docs/develop/aggregate-functions.rst:95]: It was removed because it no longer reflects the current layout of the memory.

Comment (mbasmanova) [reply velox/docs/develop/aggregate-functions.rst:95]: Would it be possible to replace it with an accurate diagram. It helps to visualize the layout.

File: velox/common/memory/AllocationPool.h (hunk 15 above)
Comment (Yuhta) [reply velox/common/memory/AllocationPool.h:40]: The change here is just for make this PR to work.  It is well contained in its own directory (`velox/common/memory`), not mixed with other changes, so you can review it separately even in the same PR.

File: velox/functions/prestosql/aggregates/DecimalAggregate.h (hunk 20 above)
Comment (Yuhta) [reply velox/functions/prestosql/aggregates/DecimalAggregate.h:211]: It's from #4154 , I can revert these and merge that PR separately after this if you prefer.

File: velox/docs/develop/aggregate-functions.rst (hunk 10 above)
Comment (Yuhta) [reply velox/docs/develop/aggregate-functions.rst:95]: The current layout is too complex to put on a diagram.  I think the new numbered list is more clear.

File: velox/functions/prestosql/aggregates/DecimalAggregate.h (hunk 20 above)
Comment (mbasmanova) [reply velox/functions/prestosql/aggregates/DecimalAggregate.h:211]: That would be helpful. Thanks.

File: velox/docs/develop/aggregate-functions.rst (hunk 10 above)
Comment (mbasmanova) [reply velox/docs/develop/aggregate-functions.rst:95]: The current layout for aggregation is the following:

- Null flags
- Free flag
- Keys
- Accumulators
- Size of variable-width accumulators

The previous diagram had null flags and accumulators. Why would adding free flag, keys and size of variable-width accumulators make it too complex?

**Comment by isadikov** [comment]
After clean/build, `AverageAggregationTest` and `SumTest` tests pass on M1.
